agent/api/task_unix.go

```diff
@@ -15,16 +15,105 @@
 
 package api
 
+import (
+	"path/filepath"
+
+	"github.com/aws/amazon-ecs-agent/agent/config"
+	"github.com/aws/amazon-ecs-agent/agent/utils"
+	"github.com/cihub/seelog"
+	docker "github.com/fsouza/go-dockerclient"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+)
+
 const (
 	portBindingHostIP = "0.0.0.0"
+	defaultCPUPeriod  = 100000 // 100ms
+	// Reference: http://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_ContainerDefinition.html
+	defaultCPUShare = 2
 )
 
 func (task *Task) adjustForPlatform() {}
 
 func getCanonicalPath(path string) string { return path }
 
-// CgroupEnabled checks whether cgroups need be enabled at the task level
-func (task *Task) CgroupEnabled() bool {
-	//stub for now
-	return false
+// BuildCgroupRoot helps build the task cgroup prefix
+// Example: /ecs/task-id
+func (task *Task) BuildCgroupRoot() (string, error) {
+	taskID, err := task.GetID()
+	if err != nil {
+		return "", errors.Wrapf(err, "task build cgroup root: unable to get task-id")
+	}
+
+	cgroupRoot := filepath.Join(config.DefaultTaskCgroupPrefix, taskID)
+	return cgroupRoot, nil
+}
+
+// BuildLinuxResourceSpec returns a linuxResources object for the task cgroup
+func (task *Task) BuildLinuxResourceSpec() (specs.LinuxResources, error) {
+	linuxResourceSpec := specs.LinuxResources{}
+
+	if !(utils.ZeroOrNil(task.VCPULimit)) {
+		taskCPUPeriod := uint64(defaultCPUPeriod)
+		taskCPUQuota := int64(task.VCPULimit * defaultCPUPeriod)
+
+		// TODO: DefaultCPUPeriod only permits 10VCPUs.
+		// Adaptive calculation of CPUPeriod required for further support
+		linuxResourceSpec.CPU = &specs.LinuxCPU{
+			Quota:  &taskCPUQuota,
+			Period: &taskCPUPeriod,
+		}
+	} else {
+		taskCPUShares := uint64(0)
+		for _, container := range task.Containers {
+			if !(utils.ZeroOrNil(container.CPU)) {
+				taskCPUShares += uint64(container.CPU)
```

Comment:
Why += and not just =?

Reply:
If you're aggregating, you should avoid the logic with the `defaultCPUShare`.  I don't think we want to include that right now.

Reply:
Could you please provide more context on why we don't want to include `defaultCPUShare` now?

Reply:
`defaultCPUShare` isn't something we schedule on and we don't make any guarantees around the weight when you assign `"cpu":0`.  If you sum `defaultCPUShare`, a task with 10 containers each with `"cpu":0` will have a weight of 20 compared to a task with 1 container that has `"cpu":0` that will end up with a weight of 2.  In each of these cases, they should have a weight of 2.

Reply:
@samuelkarp  My understanding is that summing to 20 is needed in order to preserve the current experience, even though we don't guarantee scheduling.

Reply:
When you set `"cpu":0`, you're saying "I don't want any guaranteed capacity, but I want to be able to burst."  We initially set `CpuShares` in the Docker HostConfig to 0, which got silently converted inside Docker to be 2 in the actual cgroup, because 2 is the actual minimum that the kernel enforces.  Later, a newer version of Docker started rejecting a specification of 0 and we added some compatibility code in the agent to handle that.  However, the intent here is still to get as little as possible; if you sum, you're starting to actually give a capacity guarantee (or, alternately, steal capacity under contention from containers that are *actually* intended to have a capacity guarantee).

Reply:
Hmm, so if we have a task (without task limits) but with `n containers` each without any `cpu` limits. Should we just set the task level cgroup with `CPUShares := 2`?

Reply:
Yeah, if nothing has defined limits you should set the task CPU shares to 2.

Reply:
@samuelkarp thanks for clarifying the ambiguity here